### PR TITLE
Add new `triton_fabric_vlan` data source.

### PR DESCRIPTION
**Pull request by kwilczynski**
This commit adds new data source called `triton_fabric_vlan`, which allows the
end-user to find details about a particular Fabric VLAN based on attributes such
as name, VLAN ID or description. Each of the attribute can be used as a filter
allowing to narrow down results more granularly (each filter creates an
implicit AND relationship with the other).

Additionally, this commit introduces a new `wildcardMatch` function that can be
used to perform simple pattern matches against a string value, and a function
`validateVLANIdentifier` that can be used in the schema where the `ValidateFunc`
is to be expected to verify whether the value for the Fabric VLAN ID is within
the correct range.

Signed-off-by: Krzysztof Wilczynski <kw@linux.com>

### CODE REVIEW

**Comment by kwilczynski** [comment]
To do:

- [x] Implementation
- [ ] Add documentation / code comments
- [ ] Add unit tests
- [ ] Add integration tests

File: triton/data_source_fabric_vlan.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+				Optional: true,
+				ForceNew: true,
+			},
+			"vlan_id": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateVLANIdentifier,
+			},
+			"description": {
```
Comment (stack72) [inline triton/data_source_fabric_vlan.go]: On further review, I do not believe that description is a valid search param?

File: triton/data_source_fabric_vlan.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+// dataSourceFabricVLANRead
+func dataSourceFabricVLANRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client)
+
+	vlanName, vlanNameOk := d.GetOk("name")
+	vlanID, vlanIDOk := d.GetOk("vlan_id")
+	vlanDesc, vlanDescOk := d.GetOk("description")
+
+	if !vlanNameOk && !vlanIDOk && !vlanDescOk {
+		return errors.New("")
```
Comment (stack72) [inline triton/data_source_fabric_vlan.go]: Error message

File: triton/data_source_fabric_vlan.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		return errors.Wrap(err, "error retrieving Fabric VLAN details")
+	}
+
+	matches := vlans
+
+	//
+	if vlanIDOk {
+		matches = filterVLANs(matches, func(v *network.FabricVLAN) bool {
+			return v.ID == vlanID.(int)
```
Comment (stack72) [inline triton/data_source_fabric_vlan.go]: we always return a single VLAN here?

File: triton/data_source_fabric_vlan.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		})
+	}
+	if vlanNameOk {
+		matches = filterVLANs(matches, func(v *network.FabricVLAN) bool {
+			return wildcardMatch(v.Name, vlanName.(string))
```
Comment (stack72) [inline triton/data_source_fabric_vlan.go]: we always return a single VLAN here?

File: triton/data_source_fabric_vlan.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		})
+	}
+	if vlanDescOk {
+		matches = filterVLANs(matches, func(v *network.FabricVLAN) bool {
+			return wildcardMatch(v.Description, vlanDesc.(string))
```
Comment (stack72) [inline triton/data_source_fabric_vlan.go]: we always return a single VLAN here?

File: triton/data_source_fabric_vlan.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	var vlan *network.FabricVLAN
+	if len(matches) < 1 {
+		return errors.New("")
+	}
+
+	if len(matches) > 1 {
+		log.Printf("[DEBUG] triton_fabric_vlan: Found multiple matching Fabric VLANs: %+v", matches)
+		return errors.New("")
```
Comment (stack72) [inline triton/data_source_fabric_vlan.go]: Error message?

File: triton/data_source_fabric_vlan.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	vlan = matches[0]
```
Comment (stack72) [inline triton/data_source_fabric_vlan.go]: Why reassign to a variable here rather than use matches[0] directly?

File: triton/data_source_fabric_vlan.go (hunk 2 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: Correct. This is to be added later. Still a work in progress, apologies!

File: triton/data_source_fabric_vlan.go (hunk 6 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: (same as the one above) Still a work in progress. Will add in subsequent commits.

File: triton/data_source_fabric_vlan.go (hunk 7 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: Purely readability - since it's a pointer to a struct, there is no a lot of cost (albeit, nothing is free, as a pointer also occupies some space) to have the value assigned to indicate that we want to store a single Fabric VLAN, and I prefer `vlan.Name` over `matches[0].Name`.

Here, it is more of a personal preference choice, and no problem with removing. 👍

File: triton/data_source_fabric_vlan.go (hunk 1 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: Ah, apologies! That might be the case, indeed. But, I am going to address this in a longer comment, as it is a part of the idea behind using the filters in this particular data source.

File: triton/data_source_fabric_vlan.go (hunk 6 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: Fixed.

File: triton/data_source_fabric_vlan.go (hunk 2 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: Fixed.

**Comment by kwilczynski** [comment]
As per the Pull Request [comment](https://github.com/terraform-providers/terraform-provider-triton/pull/93#discussion_r172399583), I waned to explain why do we allow description to be used as one of the possible filters in this particular data source? The answer is simple: the Fabric VLAN name is non-unique. But, more about it below...

For each of the Fabric VLANs within the same data centre, the name and description of each does not need to be unique in any way. This means, that the only element which uniquely identifies a Fabric VLAN would be its unique ID - and this in turn makes it a little bit less useful to the end-user. This is unless we allow other properties of a Fabric VLAN, namely its name and description, to be used as filters, which can be used either solely (provided that e.g. name of a Fabric VLAN is unique enough, etc.) or combined together to narrow down the number of possible matches so that there is only one possible combination of things like e.g. name and description, name and Fabric VLAN ID, etc., left which would uniquely identify a particular Fabric VLAN ID of interest. Anything that matches a single Fabric VLAN based on its ID would be an instance match for the correct Fabric VLAN of interest, but this is obvious, given that each of the Fabric VLAN IDs must be unique inside the data centre.  An example of few Fabric VLANs using Triton Public Cloud in the `us-east-1` data centre:

```json
[
  {
    "name": "My-Fabric-VLAN",
    "vlan_id": 2
  },
  {
    "description": "My Fabric VLAN",
    "name": "My-Fabric-VLAN",
    "vlan_id": 3
  },
  {
    "description": "This is a test VLAN 123",
    "name": "My-Test-VLAN",
    "vlan_id": 123
  }
]
```

As per the example above, it would be impossible for the data source to return the right Fabric VLAN ID based solely on its name, since two of the Fabric VLANs (both VLAN ID 2 and 3) share the same name.

But, when combined together with the description, then both attributes would create a more specific filter - and that would be enough identify the correct Fabric VLAN (in this case the VLAN ID would be 3). An example that could be used here in Terraform would be:

```hcl
data "triton_fabric_vlan" "main" {
  name        = "My-Fabric-VLAN"
  description = "My *"
}
```

The filters cascade one into another (provided that the attribute has been set when the data source was declared) making an implicit "AND" relationship between then. Therefore, in the example above the name and description would need both to be a match (or a best possible/longest match, if the wildcard match is used for one of them) for the right Fabric VLAN to be identified.

Basically, given that the Fabric VLAN ID name (not so much description) is non-unique within the same data centre, this ability to combine other attributes than simply `vlan_id` was the only way I could think of in terms how to make this data source somewhat useful to the end-user. Granted, this is a bit clunky... but it does work.

I hope this helps.

File: triton/data_source_fabric_vlan.go (hunk 1 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: In the [comment](https://github.com/terraform-providers/terraform-provider-triton/pull/93#issuecomment-372358777) below, I've provided a short explanation of what I found out about Fabric VLANs while working on this data source, and why I decided to use description as one of the possible filters.

File: triton/data_source_fabric_vlan.go (hunk 3 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: In a case when a Fabric VLAN ID was given, then it would always be always a correct match returning single Fabric VLAN ID, albeit we allow for results to cascade down to other filters, so that the implicit AND relationship between each is kept.

The whole logic was described in the [comment](https://github.com/terraform-providers/terraform-provider-triton/pull/93#issuecomment-372358777) below.

File: triton/data_source_fabric_vlan.go (hunk 4 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: If only name was given, then it can return multiple matching Fabric VLANs, thus we cascade down the description filter so that perhaps both would create a most strict filter resulting in only a single Fabric VLAN to be a match.

File: triton/data_source_fabric_vlan.go (hunk 5 above)
Comment (kwilczynski) [reply triton/data_source_fabric_vlan.go]: The description works in a similar fashion to the [name](https://github.com/terraform-providers/terraform-provider-triton/pull/93#discussion_r173851572) attribute.

**Comment by kwilczynski** [comment]
Unit tests:

```bash
$ go test -v ./triton/... -run TestWildcardMatch
=== RUN   TestWildcardMatch
--- PASS: TestWildcardMatch (0.00s)
PASS
ok  	github.com/kwilczynski/terraform-provider-triton/triton	0.014s
```

```bash
$ go test -v ./triton/... -run TestValidateVLANIdentifier
=== RUN   TestValidateVLANIdentifier
--- PASS: TestValidateVLANIdentifier (0.00s)
PASS
ok  	github.com/kwilczynski/terraform-provider-triton/triton	0.018s
```

**Comment by kwilczynski** [comment]
Acceptance tests:

```bash
$ make testacc TEST=./triton TESTARGS='-run TestAccTritonFabricVLAN'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./triton -v -run TestAccTritonFabricVLAN -timeout 120m
=== RUN   TestAccTritonFabricVLAN_MissingArguments
--- PASS: TestAccTritonFabricVLAN_MissingArguments (0.01s)
=== RUN   TestAccTritonFabricVLAN_BadArguments
--- PASS: TestAccTritonFabricVLAN_BadArguments (0.00s)
=== RUN   TestAccTritonFabricVLAN_Basic
--- PASS: TestAccTritonFabricVLAN_Basic (12.29s)
=== RUN   TestAccTritonFabricVLAN_WildCard
--- PASS: TestAccTritonFabricVLAN_WildCard (9.42s)
=== RUN   TestAccTritonFabricVLAN_Filters
--- PASS: TestAccTritonFabricVLAN_Filters (9.13s)
=== RUN   TestAccTritonFabricVLAN_NotFound
--- PASS: TestAccTritonFabricVLAN_NotFound (0.90s)
=== RUN   TestAccTritonFabricVLAN_FiltersNotFound
--- PASS: TestAccTritonFabricVLAN_FiltersNotFound (5.05s)
=== RUN   TestAccTritonFabricVLAN_MultipleFound
--- PASS: TestAccTritonFabricVLAN_MultipleFound (5.41s)
PASS
ok  	github.com/kwilczynski/terraform-provider-triton/triton	42.231s
```

**Comment by kwilczynski** [comment]
Added documentation:

127 0 0 1_4567_docs_providers_triton_d_triton_fabric_vlan html

**Comment by stack72** [comment]
Hi @kwilczynski 

This looks awesome - thanks for making the suggested changes :) LGTM!

P.
